test/c/test_sabre_layout.c

```diff
@@ -0,0 +1,325 @@
+        qk_circuit_gate(qc, QkGate_CX, qargs, NULL);
+    }
+    QkTranspileLayout *layout_result =
+        qk_transpiler_pass_standalone_sabre_layout(qc, target, 4, 20, 20, 2025);
+
+    QkOpCounts op_counts = qk_circuit_count_ops(qc);
+    if (op_counts.len != 2) {
+        printf("More than 2 types of gates in circuit, circuit's instructions are:\n");
+        print_circuit(qc);
+        result = EqualityError;
+        goto layout_cleanup;
+    }
+    for (int i = 0; i < op_counts.len; i++) {
+        int swap_gate = strcmp(op_counts.data[i].name, "swap");
+        int cx_gate = strcmp(op_counts.data[i].name, "cx");
+        if (cx_gate != 0 && swap_gate != 0) {
+            printf("Gate type of %s found in the circuit which isn't expected");
+            result = EqualityError;
+            goto layout_cleanup;
+        }
+        if (swap_gate == 0 && op_counts.data[i].count != 2) {
+            printf("Unexpected number of swaps %d found in the circuit.");
+            result = EqualityError;
+            goto layout_cleanup;
+        }
+    }
+    uint32_t expected_initial_layout[5] = {1, 0, 2, 3, 4};
+    uint32_t result_initial_layout[5];
+    qk_transpile_layout_initial_layout(layout_result, false, result_initial_layout);
+    for (uint32_t i = 0; i < 5; i++) {
+        if (result_initial_layout[i] != expected_initial_layout[i]) {
+            printf("Initial layout maps qubit %d to %d, expected %d instead", i,
+                   result_initial_layout[i], expected_initial_layout[i]);
+            result = EqualityError;
+            goto layout_cleanup;
+        }
+    }
+
+    uint32_t expected_final_layout[5] = {3, 0, 1, 2, 4};
+    uint32_t result_final_layout[5];
+    qk_transpile_layout_output_permutation(layout_result, result_final_layout);
+    for (uint32_t i = 0; i < 5; i++) {
+        if (result_final_layout[i] != expected_final_layout[i]) {
+            printf("Final layout maps qubit %d to %d, expected %d instead", i,
+                   result_final_layout[i], expected_final_layout[i]);
+            result = EqualityError;
+            goto layout_cleanup;
+        }
+    }
+    QkCircuit *expected_circuit = qk_circuit_new(5, 0);
+    uint32_t qargs[2] = {1, 0};
+    qk_circuit_gate(expected_circuit, QkGate_CX, qargs, NULL);
+    qargs[0] = 1;
+    qargs[1] = 2;
+    qk_circuit_gate(expected_circuit, QkGate_CX, qargs, NULL);
+    qargs[0] = 2;
+    qargs[1] = 1;
+    qk_circuit_gate(expected_circuit, QkGate_Swap, qargs, NULL);
+    qargs[0] = 2;
+    qargs[1] = 3;
+    qk_circuit_gate(expected_circuit, QkGate_CX, qargs, NULL);
+    qargs[0] = 3;
+    qargs[1] = 2;
+    qk_circuit_gate(expected_circuit, QkGate_Swap, qargs, NULL);
+    qargs[0] = 3;
+    qargs[1] = 4;
+    qk_circuit_gate(expected_circuit, QkGate_CX, qargs, NULL);
+    bool compare_result = compare_circuits(qc, expected_circuit);
+    if (!compare_result) {
+        result = EqualityError;
+        goto layout_cleanup;
+    }
```

Comment:
Mostly unimportant, but is this the standard C return-value convention for a comparator?  Things like `strcmp` and process return codes return 0 on match/success, but we're returning 0 (`false`) for the failure case.

Reply:
I can invert the logic, I didn't really think about it too much when I wrote the helper function.